azkaban-common/src/main/java/azkaban/scheduler/MissedSchedulesManager.java

```diff
@@ -0,0 +1,199 @@
+import com.google.common.annotations.VisibleForTesting;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import java.text.MessageFormat;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Miss Schedule Manager is a failure recover manager for schedules. Schedule might be missed to execute due to
+ * Trigger Scanner thread busy or web server down. This manager maintains a configurable fixed sized thread pool
+ * to execute tasks including send email notification or back execute flow.
+ * */
+@Singleton
+public class MissedSchedulesManager {
+  private static final Logger LOG = LoggerFactory.getLogger(MissedSchedulesManager.class);
+
+  private ExecutorService missedSchedulesExecutor;
+  private final boolean missedSchedulesManagerEnabled;
+  private final ProjectManager projectManager;
+  // TODO: Refactored to AlerterHolder to support missed scheduler use case
+  private final Emailer emailer;
+
+  // parameters to adjust MissScheduler
+  private final int threadPoolSize;
+  private final int DEFAULT_THREAD_POOL_SIZE = 5;
+
+  @Inject
+  public MissedSchedulesManager(final Props azkProps,
+      final ProjectManager projectManager,
+      final Emailer emailer) {
+    this.projectManager = projectManager;
+    this.emailer = emailer;
+    this.threadPoolSize = azkProps.getInt(Constants.MISSED_SCHEDULE_THREAD_POOL_SIZE, DEFAULT_THREAD_POOL_SIZE);
+    this.missedSchedulesManagerEnabled = azkProps.getBoolean(Constants.MISSED_SCHEDULE_MANAGER_ENABLED, false);
+    if (this.missedSchedulesManagerEnabled && threadPoolSize <= 0) {
+      String errorMsg = "MissedSchedulesManager is enabled but thread pool size is not positive: " + threadPoolSize;
+      LOG.error(errorMsg);
+      throw new RuntimeException(errorMsg);
+    }
+  }
+
+  /**
+   * Put timestamps for Missed executions along with email notifications into Task Queue.
+   *
+   * @param missedScheduleTimesInMs, timestamps of missed schedule
+   * @param action, execute flow action
+   * @param backExecutionEnabled, a schedule config from user, the default is false
+   * */
+  public boolean addMissedSchedule(final List<Long> missedScheduleTimesInMs,
+      final ExecuteFlowAction action,
+      final boolean backExecutionEnabled) {
+    String projectId = action.getProjectName();
+    String flowName = action.getFlowName();
+    Project project = projectManager.getProject(projectId);
+    if (project == null) {
+      return false;
+    }
+    List<String> emailRecipients = project.getFlow(flowName).getFailureEmails();
+    try {
+      Future taskFuture = this.missedSchedulesExecutor.submit(
```

Comment:
If `emailRecipients` is an empty list does it still make sense to submit a notification task?

Reply:
I guess if the user configured automatic recovery execution, but then the task to queue will be in charge of 2 things (notification + execution-triggering). Wondering if we should break it into 2 tasks. Thoughts?

Reply:
See the implementation of Emailer. If the emailRecipient List is empty there will be operations. And this sendEmail will not throw any exception if you saw the implementation, it handles and print the error message itself. So the back execution if configured will not be impacted.
But I think it would make sense to add metrics for each operations to track the failure count vs total task count.
I am planning to have separate PR to add all the necessary metrics later